### ISSUE TITLE
Using wrong library version on MacBook Pro Apple M2 Max

### ISSUE THREAD

**Issue by PeterNSteinmetz** [rpy2/rpy2#1083]
When trying to import rpy2 module or routines, the proper library is not being found and the complaint:

```OSError: cannot load library '/Library/Frameworks/R.framework/Resources/lib/libR.dylib': dlopen(/Library/Frameworks/R.framework/Resources/lib/libR.dylib, 0x0002): tried: '/Library/Frameworks/R.framework/Resources/lib/libR.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64'))```

 is generated.

I am trying to figure out how to properly fix this rather than having to install different versions of the R framework and hack symlinks. That then tends to break other things in R. 

From within Python, R_HOME is correct for arm64 as is the library path which is pointing to Library/Frameworks/R.framework/Resources/lib/libR.dylib . That is verified to contain the arm64 architecture using the file command. 

So the problem apparently resides with the library or code calling that library having been compiled as x86_64 rather than arm64. 

rpy2 v3.5.15 was installed in a venv from PyCharm. 

I understand that @lgautier may not have this platform readily available for testing, but I am willing to help here. 

What should be the next steps for trying to fix this and get a proper arm64 compilation of rpy2 c libraries working? 

Background:

OS is Sonoma 14.2.1 
R version installed is 4.3.2 platform aarch64-apple-darwin20 (64-bit)
Python 3.10

Code being run for testing:

```
import logging
import platform
import unittest

import rpy2.situation

logger = logging.getLogger(__name__)

cffi_mode = rpy2.situation.get_cffi_mode()
if cffi_mode == rpy2.situation.CFFI_MODE.API:
    import _rinterface_cffi_api as _rinterface_cffi  # type: ignore
elif cffi_mode == rpy2.situation.CFFI_MODE.ABI:
    import _rinterface_cffi_abi as _rinterface_cffi  # type: ignore
else:
    try:
        import _rinterface_cffi_api as _rinterface_cffi  # type: ignore
    except ImportError as ie_api:
        try:
            import _rinterface_cffi_abi as _rinterface_cffi  # type: ignore
        except ImportError as ie_abi:
            logger.error(f'Failed to import the API mode with "{ie_api}" '
                         'and unable to import the ABI mode.')
            raise ie_abi
ffi = _rinterface_cffi.ffi

R_HOME = rpy2.situation.get_r_home()
logging.info(f"R_HOME={R_HOME}")
lib_path = rpy2.situation.get_rlib_path(R_HOME, platform.system())
logging.info(f"lib_path={lib_path}")
rlib = ffi.dlopen(lib_path)
```

**Comment by lgautier** [comment]
Thanks for the report.

A workaround could be to force installing from the source package. This will require all needed dependencies though (C compiler, Python headers, R headers from the top of my head):
```sh
pip install --no-binary rpy2
```

I do not have a macos machine (x86_64 neither arm) to test things and rely on Github CI builds. This is where wheels are built: https://github.com/rpy2/rpy2/blob/e5656b9b454cc912e04f2d7705891b4816506028/.github/workflows/pythonpackage.yml#L69

Build artifacts for 3.5.15 are here: https://github.com/rpy2/rpy2/actions/runs/7342276602

I dabbled in using `cibuildwheel`  but at the time met too many issue. That effort is in an unfinished PR that needs to be sync'ed with the current HEAD in `master`: https://github.com/rpy2/rpy2/pull/960

**Comment by PeterNSteinmetz** [comment]
So I was able to activate the venv that I want to work in. Within that I tried `pip install --no-binary :all: rpy2`. 

This produces an error with installing the dependency cffi. Dependencies setuptools and pycparser installed ok. 

I then tried a separate installation of cffi with `pip install --no-binary :all: cffi`

and it output:

```
Collecting cffi
  Using cached cffi-1.16.0.tar.gz (512 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: pycparser in ./venv/lib/python3.10/site-packages (from cffi) (2.21)
Building wheels for collected packages: cffi
  Building wheel for cffi (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for cffi (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [51 lines of output]
      clang-6.0: error: invalid version number in 'MACOSX_DEPLOYMENT_TARGET=13'
      clang-6.0: error: invalid version number in 'MACOSX_DEPLOYMENT_TARGET=13'
      
          No working compiler found, or bogus compiler options passed to
          the compiler from Python's standard "distutils" module.  See
          the error messages above.  Likely, the problem is not related
          to CFFI but generic to the setup.py of any Python package that
          tries to compile C code.  (Hints: on OS/X 10.8, for errors about
          -mno-fused-madd see http://stackoverflow.com/questions/22313407/
          Otherwise, see https://wiki.python.org/moin/CompLangPython or
          the IRC channel #python on irc.libera.chat.)
      
          Trying to continue anyway.  If you are trying to install CFFI from
          a build done in a different context, you can ignore this warning.
      
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.macosx-13-x86_64-cpython-310
      creating build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/_imp_emulation.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/backend_ctypes.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/error.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/setuptools_ext.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/_shimmed_dist_utils.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/__init__.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/cffi_opcode.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/vengine_gen.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/pkgconfig.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/model.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/ffiplatform.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/api.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/vengine_cpy.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/commontypes.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/lock.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/recompiler.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/cparser.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/verifier.py -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/_cffi_include.h -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/parse_c_type.h -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/_embedding.h -> build/lib.macosx-13-x86_64-cpython-310/cffi
      copying src/cffi/_cffi_errors.h -> build/lib.macosx-13-x86_64-cpython-310/cffi
      running build_ext
      building '_cffi_backend' extension
      creating build/temp.macosx-13-x86_64-cpython-310
      creating build/temp.macosx-13-x86_64-cpython-310/src
      creating build/temp.macosx-13-x86_64-cpython-310/src/c
      clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.sdk -DFFI_BUILDING=1 -I/usr/local/Cellar/libffi/3.4.4/include -I/Users/peter/Documents/GitosisRepositories/TestRPy2/venv/include -I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10 -c src/c/_cffi_backend.c -o build/temp.macosx-13-x86_64-cpython-310/src/c/_cffi_backend.o -iwithsysroot/usr/include/ffi
      clang-6.0: error: invalid version number in 'MACOSX_DEPLOYMENT_TARGET=13'
      error: command '/usr/local/clang6/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for cffi
Failed to build cffi
ERROR: Could not build wheels for cffi, which is required to install pyproject.toml-based projects
```

So it seems building cffi is causing a problem. It appears that it is just trying to build for macosx-13-x86_64-cpython-310 .

Does this then require a change to the cffi project? It appears there is a similar issue at https://github.com/python-cffi/cffi/issues/41 . Or can this be fixed with some type of change in rpy2 ?

**Comment by PeterNSteinmetz** [comment]
I just also tried building cffi alone from its source distribution. A `python -m build` produced the same complaint about `clang-6.0: error: invalid version number in 'MACOSX_DEPLOYMENT_TARGET=13'`.

**Comment by lgautier** [comment]
@PeterNSteinmetz  - Does installing a binary build (wheel) for cffi work? This would make an end-to-end installation more tedious (as in 1) install cffi and dependencies using binary wheel, and 2) install rpy2 from source) but as long as it works for now...

**Comment by lgautier** [comment]
I rebased the branch trying to create binary builds using `cibuildwheel`. The artifacts produced for MacOS arm64 seem too small in size to be correct, yet the CI pipeline does not stop with an error for them:  https://github.com/rpy2/rpy2/actions/runs/7353803816

**Comment by PeterNSteinmetz** [comment]
Would that make sense if no arm64 binaries are available for cffi to incorporate? 

You likely understand these build processes better then I do, but if you look at https://github.com/python-cffi/cffi/actions/runs/7321730553/job/19942466880 it appears nothing is being built for cffi for that architecture. 

I have an issue on cffi about this at https://github.com/python-cffi/cffi/issues/44 . So far, no responses. And as I described there, I have not been able to build that binary from source on my mac due to some issues with the SDK being employed and/or some settings in their build scripts. 

From this I conclude that this will not be resolved for rpy2 until the cffi builds properly. Would you agree @lgautier ? 

The reason this problem shows up as it does is likely that the Rosetta 2 software on the mac loads and starts executing the x86 version of the cffi library. But that library then looks into the arm64 R dynamic library and does not find an x86 binary to run. 

Thus it seems like the workarounds have to be either:

1. Have an x86 version of R around and get the cffi library using it by appropriate configuration of rpy2 or modifying symlinks. For example as in https://github.com/rpy2/rpy2/issues/900 . 

2. Find a R framework version which contains libraries containing both the x86 and arm64 architectures in their dylib. 

This is a tricky sort of project isn't it? With python calling cffi and it then looking into other libraries to execute them.

**Comment by lgautier** [comment]
I fixed the PR to build wheels with `cibuildwheel` further and the binaries for Linux seem to work fine:  https://github.com/rpy2/rpy2/actions/runs/7353803816

If there is a fix that can be made at the level of `rpy2` this is the PR where that fix will be made. If the issue can only be fixed upstream (`cffi`) it will have to wait for this.

> Would that make sense if no arm64 binaries are available for cffi to incorporate?

May be. Did you try to 1) uninstall all rpy2 on your system, and then 2) install the relevant arm wheel produced by the CI pipeline?

> You likely understand these build processes better then I do, but if you look at https://github.com/python-cffi/cffi/actions/runs/7321730553/job/19942466880 it appears nothing is being built for cffi for that architecture.

My understanding is that the arm wheels are built on a self-hosted pipeline runner. What we see from the runner on GitHub just skips those build: https://github.com/python-cffi/cffi/blob/main/.github/workflows/ci.yaml#L185

> I have an issue on `cffi` about this at https://github.com/python-cffi/cffi/issues/44 . So far, no responses. And as I described there, I have not been able to build that binary from source on my mac due to some issues with the SDK being employed and/or some settings in their build scripts.

May be there are hints about how to proceed in `cffi`'s `ci.yaml` (see link above).

> From this I conclude that this will not be resolved for rpy2 until the `cffi` builds properly. Would you agree @lgautier ?

I believe that there are 3 potentials issues here:
- build `arm64` binaries for `cffi`. This is not blocking for `rpy2` since the `cffi` project does build them AFAIUI. `pip install cffi` will fetch them. Correct me if I am mistaken.
- `arm64` binaries for `rpy2` on pypi. That's what the PR #960 is trying to address.
- Build `rpy2` from source on `arm64`.

> The reason this problem shows up as it does is likely that the Rosetta 2 software on the mac loads and starts executing the x86 version of the cffi library. But that library then looks into the arm64 R dynamic library and does not find an x86 binary to run.

I am out of my depth here. I'd need to read a bit about what is happening on macos. I also vaguely remember that there are also "universal" binaries meant to run no any MacOS arch.

> Thus it seems like the workarounds have to be either:
>
>    Have an x86 version of R around and get the cffi library using it by appropriate configuration of rpy2 or modifying symlinks. For example as in 
>
>   https://github.com/rpy2/rpy2/issues/900 .

Tweaking symlinks feels like the source or hard-to-trace issues later on.

>
>    Find a R framework version which contains libraries containing both the x86 and arm64 architectures in their dylib.

At the high level it would seem seem like a solution, but I am insufficiently knowledgeable about Mac's dylibs.

> This is a tricky sort of project isn't it? With python calling cffi and it then looking into other libraries to execute them.

In that respect not more than any other Python package dynamically loading libraries on the system.

**Comment by lgautier** [comment]
I noticed that `cffi`'s support for `arm64` appeared with v1.15 (https://cffi.readthedocs.io/en/latest/whatsnew.html#v1-15-0), but `rpy2` requires `cffi` >=1.10. Do you happen to have a `cffi` >= 1.10 & < 1.15 installed?

**Comment by PeterNSteinmetz** [comment]
The normal install provides an x86 version:

```
pip install --prefer-binary cffi
Collecting cffi
  Downloading cffi-1.16.0-cp310-cp310-macosx_10_9_x86_64.whl.metadata (1.5 kB)
Requirement already satisfied: pycparser in ./lib/python3.10/site-packages (from cffi) (2.21)
Downloading cffi-1.16.0-cp310-cp310-macosx_10_9_x86_64.whl (182 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 182.5/182.5 kB 6.3 MB/s eta 0:00:00
Installing collected packages: cffi
Successfully installed cffi-1.16.0
```

To resolve the question of obtaining an appropriate arm64 binary I have been trying to pip install several versions from https://pypi.org/project/cffi/#files . None of these seem to work. 

pip install ~/Downloads/cffi-1.16.0-cp312-cp312-macosx_11_0_arm64.whl 
ERROR: cffi-1.16.0-cp312-cp312-macosx_11_0_arm64.whl is not a supported wheel on this platform.
```

```
pip install ~/Downloads/cffi-1.16.0-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl 
ERROR: cffi-1.16.0-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl is not a supported wheel on this platform.```

Thus it appears that one can get an x86 for this os but not an arm64. Which would be consistent with not being able to build it from source on this os and platform. 

Is there another one I should try or a different pip command to do a search?

**Comment by PeterNSteinmetz** [comment]
> This is not blocking for `rpy2` since the `cffi` project does build them AFAIUI. `pip install cffi` will fetch them. Correct me if I am mistaken.

I think this is the cause of the problem. The source for cffi does not appear to build an arm64 for this version of the OS or SDK. 

> I am out of my depth here. I'd need to read a bit about what is happening on macos. I also vaguely remember that there are also "universal" binaries meant to run no any MacOS arch.

Yes, one can build universal binaries but neither R nor cffi seem to be doing so. 


> Tweaking symlinks feels like the source or hard-to-trace issues later on.

It does indeed and caused me a fair amount of hassle when upgrading the OS to 10.14, which is why I would like to fix this the correct way.

**Comment by lgautier** [comment]
> The normal install provides an x86 version:
> 
> ```
> pip install --prefer-binary cffi
> Collecting cffi
>   Downloading cffi-1.16.0-cp310-cp310-macosx_10_9_x86_64.whl.metadata (1.5 kB)
> Requirement already satisfied: pycparser in ./lib/python3.10/site-packages (from cffi) (2.21)
> Downloading cffi-1.16.0-cp310-cp310-macosx_10_9_x86_64.whl (182 kB)
>    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 182.5/182.5 kB 6.3 MB/s eta 0:00:00
> Installing collected packages: cffi
> Successfully installed cffi-1.16.0
> ```
> 
> To resolve the question of obtaining an appropriate arm64 binary I have been trying to pip install several versions from https://pypi.org/project/cffi/#files . None of these seem to work.
> 
> ```
> pip install ~/Downloads/cffi-1.16.0-cp312-cp312-macosx_11_0_arm64.whl 
> ERROR: cffi-1.16.0-cp312-cp312-macosx_11_0_arm64.whl is not a supported wheel on this platform.
> ```

What is your version of MacOS? The first attempt lists MacOS 10 while the second lists MacOS 11.




> ```
> pip install ~/Downloads/cffi-1.16.0-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl 
> ERROR: cffi-1.16.0-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl is not a supported wheel on this platform.
> ```
> 
> Thus it appears that one can get an x86 for this os but not an arm64. Which would be consistent with not being able to build it from source on this os and platform.
> 
> Is there another one I should try or a different pip command to do a search?

Not to my knowledge. I'll share if I read about something else to try.

**Comment by lgautier** [comment]
https://pypi.org/project/cffi/#files seems to only offer arm64 binary for Mac OS 10 or 11. If you have 12, or 13, you'll have to install cffi from source.

**Comment by lgautier** [comment]
IIUIC this is pointing toward an issue upstream: install cffi. The first step should be to manage to install cffi from source on your MacOS (presumably) > 11.

There are few pages found when search about this specific issue. For example, 
https://stackoverflow.com/questions/73247740/cffi-installation-got-failure-on-macos-12-5-with-pip

I can't help further unfortunately.

**Comment by PeterNSteinmetz** [comment]
@lgautier Thanks for looking into it. I will continue to track and work on it a bit with the cffi folks and report back if anything changes.

**Comment by PeterNSteinmetz** [comment]
So I found a hint to what might be going on in the pytables project at https://github.com/PyTables/PyTables/issues/887 and more information about where the Apple binaries are supposed to be at https://stackoverflow.com/questions/64992467/mac-clang-installation-seems-to-override-gcc-install . 

It appears that some older version of homebrew or another project required installing a clang6 version at /usr/local/clang6 and having that earlier in your PATH. Thus clang was invoking an older version of the compiler which did not understand 'MACOSX_DEPLOYMENT_TARGET=14'.

I finally realized what was going on when I invoked `clang6 --version` and noticed it was giving me an x86 compiler with an older version. So that is the thing to check if you are having a similar problem. 

Once the paths were set correctly and the correct clang was being used, I was able to build rpy2 from source.
